mozilla/kuma
# bug 1525719: publish/unpublish to/from the doc API

#### Description
This PR provides the foundation for connecting, via the document API, the Wiki side of MDN (creating, editing, deleting, restoring, reverting, and moving documents) with the new read-only, `beta` domain side of MDN. All operations on the Wiki side (listed above) will be automatically reflected to the document API via handlers connected to Django signals (i.e., `post_delete`, `render_done`, and `restore_done`). Those three signals, according to my testing, cover all of the operations listed above, triggering handlers that in turn either `publish` or `unpublish` to the S3 bucket that serves the document API. So for example, when a document is deleted, the `post_delete` handler will launch an `unpublish` task into Celery, which when run within a Celery worker, will delete the pre-rendered JSON file in S3 that corresponds to that document.

This PR also provides two new management commands: `publish` and `unpublish`. The purpose of both of these commands is twofold. Primarily, they provide a way to manually `publish` or `unpublish` for those times when the automatic Celery tasks fail for whatever reason. However, in the case of the `publish` command, it also exists to "prime the pump" so to speak, to initially publish all of the documents.

The idea here is that, after an initial publication of all of the documents, the document API will automatically track all changes as they're made within the Wiki domain: document re-renders, document edits, document reverts, document restores, document moves, the creation of new documents, and so on.

The `publish` and `unpublish` Celery tasks are brokered through the new `mdn_api` Celery queue. Initially, all of the Celery workers will be configured to pull from that queue (see https://github.com/mdn/infra/pull/236), but I've created that new queue so that in the near future those tasks could be handled by one or more Celery workers configured to use a read-replica of the current production MySQL RDS instance in order to avoid loading-down the production database.

#### Infrastructure Context
The infrastructure context here is that the document API will be served via CloudFront backed by an S3 origin serving pre-rendered JSON files (and handling redirects from one key in S3 to another). This provides a simple, fast, and efficient way to serve the document API, and is already in place within the staging environment (for now the TTL on the stage CloudFront instance for all `/api/v1/doc/*` paths is 12 hours). There is no need for invalidation of paths within CloudFront. When the TTL expires (after 12 hours or whatever we finally choose) for any given path, CloudFront will return to S3 to get the latest data (if needed, since it'll make a conditional request using the `ETag` and `If-None-Match` HTTP headers). Of course, this means that any document page within the read-only `beta` domain may be at most 12 hours out-of-date with the same document page in the Wiki domain.

#### Redirects
This document API as implemented within this PR also handles content-based redirects (a redirect directive embedded within a document). There are two kinds of content-based redirects (both are 301's): a redirect from one document to another, as well as a redirect from a document to a non-document page. The first kind of redirect, the redirect from one document to another, has been built upon S3's ability to handle redirects. The origin for CloudFront uses the S3 bucket's "website" domain rather than its raw domain, since the "website" domain will automatically handle the redirection locations embedded as metadata within the S3 objects. So, for example, a request via the document API for `/api/v1/doc/en-US/MDN/Promote` will be 301 re-directed to `/api/v1/doc/en-US/MDN/About/Promote` (this 301 is cached within the CDN after the first miss), so the `fetch` within the React `DocumentProvider` call must be configured to follow redirects. Since the returned JSON, in the case of redirects from one document to another, will be for a different locale and/or slug, the `DocumentProvider` code must account for that and update the `history` accordingly (which I've taken a stab at within this PR, but only as a first attempt, and I'll leave it to @davidflanagan to finalize that code).

The redirects from document pages to non-document pages are handled by passing the URL for the non-document page in the `redirectURL` attribute of the JSON returned by the document API, which is normally `null`. When it's not `null`, it contains either a relative or full URL to a non-document page.

#### Testing
I've tested this locally, and in fact populated the stage S3 bucket serving the document API from my sample database. The AWS access keys are already deployed as secrets within the stage, production, and standby Kubernetes clusters, and the stage/production S3 bucket names are already baked into the configuration settings for stage, production, and standby (for example. see https://github.com/mdn/infra/blob/ad561b1fb13852ac2564334b71f80ed5d9ff307f/apps/mdn/mdn-aws/k8s/regions/oregon/stage.sh#L106). Once this PR is merged, I'll re-publish the document API for stage from the actual stage database, but for now there are enough documents to test on stage.

Even though you won't have the AWS credentials for actually publishing/unpublishing to/from S3, i'd still recommend confirming that those tasks are launched (you'll see either a `Skipping publish for XXX: no S3 bucket configured` or `Skipping unpublish for XXX: no S3 bucket configured` message in the `worker` log -- `docker-compose logs -f worker`) when you try each of the following:
- create a new document
- edit a document
- delete a document
- restore a deleted document
- revert a document to an earlier revision
- move a document

You can also do this while testing the management commands, for example:
```sh
docker-compose exec web bash
./manage.py publish --locale en-US --chunk-size 10
./manage.py unpublish /en-US/docs/Web/CSS /en-US/docs/Web/HTML
```
